plugins/hls-gadt-plugin/src/Ide/Plugin/GADT.hs

```diff
@@ -0,0 +1,135 @@
+    }
+
+-- | Parameter used in the command
+data ToGADTParams = ToGADTParams
+    { uri          :: Uri
+    , range        :: Range
+    -- PluginId doesn't have instance of ToJSON and FromJSON
+    , pluginIdText :: T.Text
+    } deriving (Generic, ToJSON, FromJSON)
+
+toGADTSyntaxCommandId :: CommandId
+toGADTSyntaxCommandId = "GADT.toGADT"
+
+-- | A command replaces H98 data decl with GADT decl in place
+toGADTCommand :: CommandFunction IdeState ToGADTParams
+toGADTCommand state ToGADTParams{..} = response $ do
+    let plId = PluginId pluginIdText
+    nfp <- getNormalizedFilePath plId uri
+    (decls, exts) <- getInRangeH98DeclsAndExts state range nfp
+    (L ann decl) <- handleMaybe
+        ("Expect 1 decl, but got " <> show (Prelude.length decls))
+        (if Prelude.length decls == 1 then Just $ head decls else Nothing)
+    deps <- liftIO $ runAction "GADT.GhcSessionDeps" state $ use GhcSessionDeps nfp
+    (hsc_dflags . hscEnv -> df) <- ExceptT
+        $ pure
+        $ maybeToEither "Get GhcSessionDeps failed" deps
+    txt <- ExceptT $ pure $ T.pack <$> (prettyGADTDecl df . h98ToGADTDecl) decl
+    range <- ExceptT
+        $ pure
+        $ maybeToEither "Unable to get data decl range"
+        $ srcSpanToRange $ locA ann
+    pragma <- getNextPragma state nfp
+    let insertEdit = [insertNewPragma pragma GADTs | all (`notElem` exts) [GADTSyntax, GADTs]]
+
+    _ <- lift $ sendRequest
+            SWorkspaceApplyEdit
+            (ApplyWorkspaceEditParams Nothing (workSpaceEdit nfp (TextEdit range txt : insertEdit)))
+            (\_ -> pure ())
+
+    pure Null
+    where
+        workSpaceEdit nfp edits = WorkspaceEdit
+            (pure $ HashMap.fromList
+                [(filePathToUri $ fromNormalizedFilePath nfp,
+                 List edits)])
+                 Nothing Nothing
+
+codeActionHandler :: PluginMethodHandler IdeState TextDocumentCodeAction
+codeActionHandler state plId@(PluginId txt) (CodeActionParams _ _ doc range _) = response $ do
+    nfp <- getNormalizedFilePath plId doc
+    (inRangeH98Decls, _) <- getInRangeH98DeclsAndExts state range nfp
+    let actions = map (mkAction . printOutputable . tcdLName . unLoc) inRangeH98Decls
+    pure $ List actions
+    where
+        mkAction :: T.Text -> Command |? CodeAction
+        mkAction name = InR CodeAction{..}
+            where
+                _title = "Convert \"" <> name <> "\" to GADT syntax"
+                _kind = Just CodeActionRefactorRewrite
+                _diagnostics = Nothing
+                _isPreferred = Nothing
+                _disabled = Nothing
+                _edit = Nothing
+                _command = Just
+                    $ mkLspCommand plId toGADTSyntaxCommandId _title (Just [toJSON mkParam])
+                _xdata = Nothing
+
+        mkParam = ToGADTParams (doc ^. L.uri) range txt
+
+-- | Get all H98 decls in the given range, and enabled extensions
+getInRangeH98DeclsAndExts :: (Monad (t IO), MonadTrans t) =>
```

Comment:
Generally I think you could make your life a bit easier with some `mtl`: you could have `(MonadError String m, MonadIO m)` here and just use `mtl` functions instead of lots of explicit lifting. I think that's true throughout.

Reply:
Thank you for pointing it out! I'm not familiar with mtl indeed.